x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/RuleQueryBuilder.java

```diff
@@ -174,34 +174,29 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
 
     @Override
     protected Query doToQuery(SearchExecutionContext context) throws IOException {
-        if ((pinnedIds != null && pinnedIds.isEmpty() == false) && (pinnedDocs != null && pinnedDocs.isEmpty() == false)) {
-            throw new IllegalArgumentException("applied rules contain both pinned ids and pinned docs, only one of ids or docs is allowed");
-        }
-
-        if (pinnedIds != null && pinnedIds.isEmpty() == false) {
-            PinnedQueryBuilder pinnedQueryBuilder = new PinnedQueryBuilder(organicQuery, pinnedIds.toArray(new String[0]));
-            return pinnedQueryBuilder.toQuery(context);
-        } else if (pinnedDocs != null && pinnedDocs.isEmpty() == false) {
-            PinnedQueryBuilder pinnedQueryBuilder = new PinnedQueryBuilder(organicQuery, pinnedDocs.toArray(new Item[0]));
-            return pinnedQueryBuilder.toQuery(context);
-        } else {
-            return organicQuery.toQuery(context);
-        }
-
+        throw new IllegalStateException(NAME + " should have been rewritten to another query type");
     }
 
     @SuppressWarnings("unchecked")
     @Override
     protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
-        if (pinnedIds != null || pinnedDocs != null) {
-            return this;
+        if ((pinnedIds != null && pinnedIds.isEmpty() == false) && (pinnedDocs != null && pinnedDocs.isEmpty() == false)) {
+            throw new IllegalArgumentException("applied rules contain both pinned ids and pinned docs, only one of ids or docs is allowed");
+        }
+
+        if (pinnedIds != null) {
```

Comment:
There are several 'else if' statements here. Since all the branches of this nested **if..else** include a return statement, removing the 'else if' statements would enhance readability for developers.

```java

        if (pinnedIds != null) {
            return new PinnedQueryBuilder(organicQuery, pinnedIds.toArray(new String[0]));
        }

        if (pinnedDocs != null) {
            return new PinnedQueryBuilder(organicQuery, pinnedDocs.toArray(new Item[0]));
        }
        
        if (pinnedIdsSupplier != null || pinnedDocsSupplier != null) {
            List<String> identifiedPinnedIds = pinnedIdsSupplier != null ? pinnedIdsSupplier.get() : null;
            List<Item> identifiedPinnedDocs = pinnedDocsSupplier != null ? pinnedDocsSupplier.get() : null;
            if (identifiedPinnedIds != null && identifiedPinnedIds.isEmpty() == false) {
                return new PinnedQueryBuilder(organicQuery, identifiedPinnedIds.toArray(new String[0]));
            }
            if (identifiedPinnedDocs != null) {
                return new PinnedQueryBuilder(organicQuery, identifiedPinnedDocs.toArray(new Item[0]));
            }
            return this;
        }
```

Reply:
I have a followup PR to remove pinnedIds altogether which will greatly simplify this.